core/src/main/kotlin/org/evomaster/core/search/service/monitor/SearchProcessMonitor.kt

```diff
@@ -0,0 +1,189 @@
+    var step :StepOfSearchProcess<*>? = null
+    var isMutated : Boolean = false
+
+    private var cs = ""
+    private var tb = 1
+
+    private val evaluatedIndividuals : MutableList<EvaluatedIndividual<*>> = mutableListOf()
+
+    companion object {
+        var doesSave = true
+        val DATA_FOLDER = "data"
+        val gson =GsonBuilder().registerTypeAdapter(RestAction::class.java, InterfaceAdapter<RestAction>())
+                .registerTypeAdapter(Param::class.java, InterfaceAdapter<Param>())
+                .registerTypeAdapter(Gene::class.java, InterfaceAdapter<Gene>())
+                .create()
+
+        //FIXME need to change accordingly
+        var digit = 7
+
+        fun getFileType() : String{
+            return ".json"
+        }
+
+        fun getInt(digit:Int, value : Int) :String{
+            return String.format("%0$digit"+"d", value)
+        }
+
+        fun covertInt(digit: Int, value : String) : Int{
+            var s = value.indexOfFirst { it -> it.toInt() != 0 }
+            return value.substring(s).toInt()
+        }
+
+        fun getOverallName() : String{
+            return "overall"
+        }
+
+//        fun generateAlgoFolder(algo : String, stoppingCriterion: String, sampling: String, budget: Int, maxTestSize: Int, population : Int,  indexOfRun : Int= -1): String{
+//            return algo + "_"+stoppingCriterion+"_"+sampling+"_"+budget + "_"+ maxTestSize+ (if(population != 30) "_P"+population.toString() else "")+ (if(indexOfRun != -1) "_R"+indexOfRun.toString() else "")
+//        }
+
+//        fun generateAlgoFolder (_config : EMConfig, indexOfRun: Int): String{
+//            return generateAlgoFolder (_config.algorithm.name, _config.stoppingCriterion.toString(), _config.smartSampling.toString(),
+//                    (if(_config.stoppingCriterion.toString().toLowerCase().contains("time")) _config.maxTimeInSeconds else _config.maxActionEvaluations), _config.maxTestSize, _config.populationSize, indexOfRun)
+//
+//        }
+    }
+
+    @PostConstruct
+    fun postConstruct(){
+        // TODO configure monitoring the details of search process
+        if(config.enableProcessMonitor){
+            time.addListener(this)
+
+            setCS(config.statisticsColumnId)
+            initMonitorProcessOuputs()
+        }
+    }
+
+    override fun newActionEvaluated() {
+        if(config.enableProcessMonitor){
+            evaluatedIndividuals.add(eval!!)
+            step = StepOfSearchProcess(archive, time.evaluatedIndividuals, eval!!.individual, eval!!, System.currentTimeMillis(),isMutated)
+        }
+
+    }
+
+    //TODO Man
+    fun record(added: Boolean, improveArchive : Boolean, evalInd : EvaluatedIndividual<*>){
+        if(config.enableProcessMonitor){
+            if(evalInd != eval) throw IllegalStateException("Mismatched evaluated individual under monitor")
+            if(doesSave){
+                if(time.evaluatedActions > tb * 100){
+                    step!!.added = added
```

Comment:
why is "step" nullable and guaranted not nullable here? add comment

Reply:
Yes, the comment is added.